### PR TITLE
ACA-85: Check for NoC Approval

### PR DESCRIPTION
**Pull request by smathangi**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/ACA-85

### Change description ###

Check for NoC Approval Endpoint.
 
LLD : https://tools.hmcts.net/confluence/display/ACA/API+Operation%3A+Check+for+Notice+of+Change+Approval

Related data-store PR : https://github.com/hmcts/ccd-data-store-api/pull/1207

IDAM tickets for new system user account

https://tools.hmcts.net/jira/browse/SIDM-5164
https://tools.hmcts.net/jira/browse/RDO-9670

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[ x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/managecase/repository/DataStoreRepository.java
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/reform/managecase/repository/DataStoreRepository.java:31]: > CaseResource findCaseByCaseIdExternalApi(String caseId);
though this api doesn't need for approver.. probably worth to inline with other api signatures??

Alternatively define two specific api's wherever required and abstract userToken population inside repo.

CaseViewResource findCaseByCaseIdForCaa(String caseId);
CaseViewResource findCaseByCaseIdForApprover(String caseId);

File: src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java:317]: any possibility to enforce these as Jsr303 hibernate annotations on ChangeOrganisationRequest.

File: src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java:306]: 200 in this case?

Comment (tomxelliott) [reply src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java:306]: Yeah I checked with Matt and he said that we quietly return OK because it isn't an error

File: src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java
Comment (tomxelliott) [reply src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java:317]: I am sure I can, probably will be much cleaner 👍

File: src/main/java/uk/gov/hmcts/reform/managecase/service/NoticeOfChangeService.java
Comment (smathangi) [inline src/main/java/uk/gov/hmcts/reform/managecase/service/NoticeOfChangeService.java:313]: you need to inject NocApprovalDataStoreRepository in this case.

Comment (tomxelliott) [reply src/main/java/uk/gov/hmcts/reform/managecase/service/NoticeOfChangeService.java:313]: Please elaborate...

File: src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/ChangeOrganisationRequest.java
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/ChangeOrganisationRequest.java:40]: Minor, but does this need some unit testing?

File: src/main/java/uk/gov/hmcts/reform/managecase/ApplicationParams.java
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/reform/managecase/ApplicationParams.java:13]: Kind of consolidated class for most of the injected app params .

File: src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/model/CaseViewActionableEvent.java
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/model/CaseViewActionableEvent.java:8]: In general I prefer duplicate classes on each side (to inline with json contract) rather than shared models.

File: src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java:64]: Should this be capitalised for consistency ?

`Auto approval is not configured for the case`

File: src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java:65]: `Check Notice of Change approval has completed`

or

`Check Notice of Change approval has finished`

File: src/functionalTest/resources/features/common/mca/Check_NoC_Approval__Base.td.json
Comment (mattnayler) [inline src/functionalTest/resources/features/common/mca/Check_NoC_Approval__Base.td.json:10]: If mimicking data-store calling your callback: should `s2sClientId` not be `"ccd_data"`?

NB: you may need to add `ccd_data` equivalents to the following to make this work:
* befta env vars: https://github.com/hmcts/aac-manage-case-assignment/blob/ACA-85/aca-docker/bin/env_variables_all.txt#L6-L7
* jenkins updates:
   * S2S Key: https://github.com/hmcts/aac-manage-case-assignment/blob/ACA-85/Jenkinsfile_CNP#L39
   * S2S ID: https://github.com/hmcts/aac-manage-case-assignment/blob/ACA-85/Jenkinsfile_CNP#L66

NB: Although updates to S2S key in docker config is not required: https://github.com/hmcts/ccd-docker/blob/master/compose/backend.yml#L122

File: src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/StartEventResource.java
Comment (mattnayler) [inline src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/StartEventResource.java:15]: If we really need two forms of CaseDetails classes: can we use a different name as not all cases use callbacks.

File: src/main/java/uk/gov/hmcts/reform/managecase/service/NoticeOfChangeApprovalService.java
Comment (mattnayler) [inline src/main/java/uk/gov/hmcts/reform/managecase/service/NoticeOfChangeApprovalService.java:1]: Should this be moved in to the child NOC package?

File: src/test/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeControllerTest.java
Comment (mattnayler) [inline src/test/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeControllerTest.java:469]: Please also verify that the correct down stream service `noticeOfChangeApprovalService` has been called for happy paths that are approved.  **and in the non-approved test verify that it has not been called.**

**Review by mattnayler — CHANGES_REQUESTED** [review]

File: src/functionalTest/resources/features/F-207 - Check Notice Of Change Approval/F-207 - Check Notice Of Change Approval.feature
Comment (smathangi) [reply src/functionalTest/resources/features/F-207 - Check Notice Of Change Approval/F-207 - Check Notice Of Change Approval.feature:22]: Correct. Thank you.

File: src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/ChangeOrganisationRequest.java
Comment (tomxelliott) [reply src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/ChangeOrganisationRequest.java:45]: The LLD states that all must be present in the COR so I set up the check to make sure all fields are present.

File: src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/CaseDetails.java
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/CaseDetails.java:35]: Yes, that looks like a hack to me to satisfy multiple versions (ES and rest api)

File: src/functionalTest/resources/features/common/mca/Check_NoC_Approval__Base.td.json
Comment (smathangi) [reply src/functionalTest/resources/features/common/mca/Check_NoC_Approval__Base.td.json:10]: Ideally yes but MCA endpoint code doesn't care whether it's ccd_data or xui_webapp.

Comment (mattnayler) [reply src/functionalTest/resources/features/common/mca/Check_NoC_Approval__Base.td.json:10]: Agreed but using the correct code will allow us to test that your `MANAGE_CASE_S2S_AUTHORISED_SERVICES` change has been preserved and applied to the MCA instance.

File: Jenkinsfile_CNP
Comment (tomxelliott) [inline Jenkinsfile_CNP:24]: Is this a typo?

Comment (smathangi) [reply Jenkinsfile_CNP:24]: yes, copy from search command 😄

File: src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/ChangeOrganisationRequest.java
Comment (smathangi) [reply src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/ChangeOrganisationRequest.java:40]: done

File: src/integrationTest/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeControllerIT.java
Comment (mattnayler) [reply src/integrationTest/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeControllerIT.java]: Nice use of text fixtures to resolve this issue.

File: src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java]: Alternative wording now applied in as an update for https://github.com/hmcts/aac-manage-case-assignment/pull/129#discussion_r521519161

File: src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java]: Alternative wording now applied in as an update for https://github.com/hmcts/aac-manage-case-assignment/pull/129#discussion_r521519161

File: src/functionalTest/resources/features/F-207 - Check Notice Of Change Approval/S-207.3.td.json
Comment (mattnayler) [inline src/functionalTest/resources/features/F-207 - Check Notice Of Change Approval/S-207.3.td.json:16]: Should this not be `ACA_Users_Caseworker_CAA`?

File: src/functionalTest/resources/features/F-207 - Check Notice Of Change Approval/S-207.4.td.json
Comment (mattnayler) [inline src/functionalTest/resources/features/F-207 - Check Notice Of Change Approval/S-207.4.td.json:16]: Should this not be `ACA_Users_Caseworker_CAA`?

File: src/functionalTest/resources/features/F-207 - Check Notice Of Change Approval/S-207.5.td.json
Comment (mattnayler) [inline src/functionalTest/resources/features/F-207 - Check Notice Of Change Approval/S-207.5.td.json:16]: Should this not be `ACA_Users_Caseworker_CAA`?

File: src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/CallbackCaseDetails.java
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/CallbackCaseDetails.java]: Resolved in updated `CaseDetails` class.

**Review by mattnayler — APPROVED** [review]
LGTM 👍.  However 4 minor adjustments requested on tests.

File: src/integrationTest/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeControllerIT.java
Comment (sullivangareth) [reply src/integrationTest/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeControllerIT.java:431]: This has been introduced in ACA-68, but only has APPROVED or PENDING

File: src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java
Comment (sullivangareth) [reply src/main/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeController.java]: Enum has been introduced with values for PENDING and APPROVED in ACA-68

File: src/main/java/uk/gov/hmcts/reform/managecase/repository/DefaultDataStoreRepository.java
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/managecase/repository/DefaultDataStoreRepository.java:134]: Can this be renamed to remove the word "Only"

`submitEventForCase` will be renamed as part of code review changes for ACA-68 in #150

File: src/main/resources/application.yaml
Comment (mattnayler) [inline src/main/resources/application.yaml:51]: Typo in environment variable reference: `IDAM_CAA_PASSWORD`

File: src/integrationTest/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeControllerIT.java
Comment (mattnayler) [reply src/integrationTest/java/uk/gov/hmcts/reform/managecase/api/controller/NoticeOfChangeControllerIT.java:430]: Could this be updated to use the `PENDING` enum as this is the default status after NoC Request has been raised when auto-approval is not configured?

File: src/test/java/uk/gov/hmcts/reform/managecase/client/datastore/ChangeOrganisationRequestTest.java
Comment (mattnayler) [inline src/test/java/uk/gov/hmcts/reform/managecase/client/datastore/ChangeOrganisationRequestTest.java:1]: Move to the package that corresponds to the model under test: i.e. `uk.gov.hmcts.reform.managecase.domain`.

File: src/main/java/uk/gov/hmcts/reform/managecase/repository/DefaultDataStoreRepository.java
Comment (mattnayler) [inline src/main/java/uk/gov/hmcts/reform/managecase/repository/DefaultDataStoreRepository.java:133]: Can this be renamed to remove the word "Only"?

NB: The previous `submitEventForCase` has be renamed as part of code review changes for ACA-68 in #150.

**Review by mattnayler — CHANGES_REQUESTED** [review]
Some minor review comments add or re-raised as still waiting a response or action.

File: src/main/java/uk/gov/hmcts/reform/managecase/repository/DefaultDataStoreRepository.java
Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/reform/managecase/repository/DefaultDataStoreRepository.java]: [re-raised](#discussion_r533373203) during a re-review of current code.  So resolving this earlier conversation.

File: src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/DataStoreApiClient.java
Comment (sullivangareth) [inline src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/DataStoreApiClient.java:37]: This results in URL with too many slashes - `internal//cases/{caseId}/event-triggers/{eventId}`

Change to 

`String START_EVENT_TRIGGER = INTERNAL_CASES + "/event-triggers/{eventId}";`

Comment (mattnayler) [reply src/main/java/uk/gov/hmcts/reform/managecase/client/datastore/DataStoreApiClient.java]: Seen resolved in 1021e05.

**Review by mattnayler — APPROVED** [review]
LGTM 👍

**Review by tomxelliott — APPROVED** [review]
All looks good to me 👍

**Pull request merged**
